### PR TITLE
Enable DCE on OSSA

### PR DESCRIPTION
**Pull request by meg-gupta**
- Support destroy_value/copy_value/begin_borrow/end_borrow
- Support adding reverse dependencies for SILArgument
- Insert lifetime end instructions (destroy_value/end_borrow) while deleting instructions which would have ended the lifetime of its operands

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci benchmark

File: lib/SILOptimizer/Transforms/DeadCodeElimination.cpp [hunk 1]
```diff
@@ -94,15 +107,16 @@ class DCE : public SILFunctionTransform {
...
+  // the dependency goes from inst_b to inst_a: if inst_b is alive then
+  // inst_a is also alive.
   // For some instructions the dependency is exactly the other way round, e.g.
   //   %1 = inst_which_can_fail
   //   cond_fail(%1)
   // In this case cond_fail is alive only if inst_which_can_fail is alive.
   // The key of this map is the source of the dependency (inst_a), the
-  // value is the destination (inst_b).
-  llvm::DenseMap<SILInstruction *, SILInstruction *> ReverseDependencies;
+  // value is the set of instructions dependent on it (inst_b).
```
Comment (eeckstein) [inline lib/SILOptimizer/Transforms/DeadCodeElimination.cpp:117]: It makes sense to let multiple instructions be the destination of a reverse dependency.
Did you add a test case for this?

**Review by eeckstein — APPROVED** [review]
LGTM!

File: lib/SILOptimizer/Transforms/DeadCodeElimination.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/DeadCodeElimination.cpp:117]: There were some tests already with multiple `fix_lifetime`. I added an additional test with multiple `destroy_value` now.

**Comment by meg-gupta** [comment]
Thanks for the review @eeckstein

**Comment by meg-gupta** [comment]
@swift-ci test and merge

**Pull request merged**